README.md
# Bitcoin Vault Api

A simple public JSON API written in Python and Flask. Includes a proposed production deployment system using Docker, NGINX, uWSGI and MongoDB. Deployed under http://api.bitcoinvault.global

The API consists of 4 parts closed in separate Docker containers: the NGINX server redirecting queries to the Flask application, the Flask application executing incoming queries, BTCV node and a database that stores blockchain statistics along with a daemon updating the data.

# NGINX and Flask app

Both components work synergistically to handle requests coming from outside. NGINX is a kind of avant-garde when it comes to handling incoming queries. Sending the request to the API takes place through the handling of the http protocol (port 8080) to the appropriate address on which the service was hosted. Then, using the uWSGI protocol, the query is forwarded to the Flask application running on port 5000. By connecting to the database, the application handles user queries and returns the required results.

At the moment, there are 6 queries that the user can request::

```python
/miningdifficulty
```

The average mining difficulty query for x consecutive blocks in the last y blocks, where x and y are counted based on the given period. The request has one parameter `period`, which can take three values: `week`, `month`, `all` (every other value than `week` and `month`).

`week` (default): average mining difficulty is calculated every 6 consecutive blocks for the last 1008 blocks of blockchain,

`month`: average mining difficulty is calculated every 144 consecutive blocks for the last 4320 blocks of blockchain,

`all`: average mining difficulty is calculated every 1008 consecuitve blocks for all blocks of blockchain.
    
```python
/transactionvolume
```

The average transaction volume query for x consecutive blocks in the last y blocks, where x and y are counted based on the given period. The request has one parameter `period`, which can take three values: `week`, `month`, `all` (every other value than `week` and `month`).

`week` (default): average transaction volume is calculated every 6 consecutive blocks for the last 1008 blocks of blockchain,

`month`: average transaction volume is calculated every 144 consecutive blocks for the last 4320 blocks of blockchain,

`all`: average transaction volume is calculated every 1008 consecuitve blocks for all blocks of blockchain.
    
```python
/nonemptywalletsnumber
```

The query returns the number of current addresses where the BTCV coin balance is greater than 0.

```python
/piechartdata
```

The query returns the address distribution for coinbase transactions from the last 100 blocks.

```python
/richestwallets
```

The query returns a list of 20 addresses with the most BTCV coins at the current moment, along with their balance.

```python
/status
```

The query returns the current blockchain height, network hashrate and the number of all coins currently in circulation.

# Bitcoin Vault node

This component gets the current version of the BTCV node from the Github repository: `https://github.com/bitcoinvault/bitcoinvault`.
The task of this component is to synchronize the BTCV blockchain so that the data on which the analysis is performed is up-to-date.

# MongoDB

This component runs the Mongo database in which blockchain statistics necessary to execute queries for the API are stored. The component actually consists of two parts: the Mongo database and the daemon, which updates this database with data by processing the blockchain. Daemon queries the BTCV node about the blockchain status every 5 seconds and updates the statistics if the database status does not match the blockchain status.

The differences between the state of the base and the state of the blockchain may be due to two reasons: the database does not have statistics from all blocks available in the blockchain (the number of processed blocks for statistics is smaller than the actual number of blocks in the chain) or the reorg occured in blockchain. Both situations are properly handled: the first - by processing the missing blocks and updating the database state, the second - by restoring the last state of the database before the reorg occurence and updating the database from there to the current state of the blockchain. 

Restoring the database is possible thanks to periodic database backups, made every full thousandth block. The database backup files are saved in `/var/backups/mongodb`. For creating and restoring database daemon uses `mongodump` and `mongorestore` commands provided by MongoDB.

The daemon "talks" to the BTCV node using the RPC methods provided by the node.
RPC server parameters are provided in `rpc.py` file:

```python
url = "http://user:password@bitcoind:8332"
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.4 (2021-12-16)
### Fix
* Remove assigning txids to addresses to avoid exceeding document size. ([`81edd47`](https://github.com/bitcoinvault/api/commit/81edd479c0205b0996e8f0318bb34db2326f1e7d))

## v0.1.3 (2021-06-24)
### Fix
* **tx:** Add missing fields for Tx model ([`4e58ff6`](https://github.com/bitcoinvault/api/commit/4e58ff6a86960bc08a78a3e8f9f0edef570355d7))

## v0.1.2 (2021-06-01)
### Performance
* **dockerfile:** Use alpine image as base ([`39814b5`](https://github.com/bitcoinvault/api/commit/39814b5c5a87c28c196c2bd88b07313f819388d8))

## v0.1.1 (2021-03-19)
### Fix
* **richestwallets:** Fix possibility to return more than 20 of richest wallets ([#11](https://github.com/bitcoinvault/api/issues/11)) ([`8482a14`](https://github.com/bitcoinvault/api/commit/8482a146321f405a10b05fb923433f8be8873004))

## v0.1.0 (2021-02-26)
### Feature
* **docker:** Update version of Python to 3.8 ([`0febd0b`](https://github.com/bitcoinvault/api/commit/0febd0b0cb3c76b35145b4043db3557d3b51387d))

## v0.0.2 (2021-02-24)
### Fix
* **ci:** Add PAT to be able to create new release ([`3d6baec`](https://github.com/bitcoinvault/api/commit/3d6baecc49248e914fcb4c631c2dc3c51f3386a7))
* **ci:** Add PAT to be able to create new release ([`0b6306b`](https://github.com/bitcoinvault/api/commit/0b6306b38fc4dad5df2b98efcb81b81a21fff439))
* Proper imports ([`2717123`](https://github.com/bitcoinvault/api/commit/2717123705f38571d589b2a6ed94301db2d0356c))

## v0.0.1 (2021-02-24)
### Fix
* Copy src after installing dependencies, to use layer caching ([`593433a`](https://github.com/bitcoinvault/api/commit/593433a9f14f4b09635c7e76faa7db0d54cd9a80))
* Remove backup, fix imports ([`09e17f5`](https://github.com/bitcoinvault/api/commit/09e17f517078f1caaafd4a3e53430aab9a27356a))


src/api.py
from flask import Flask, jsonify, request
from flask_mongoengine import MongoEngine

from blockchain.blockchain_analyzer import BlockchainAnalyzer
from db.db import get_addresses, get_blockchain, get_utxos, get_db_uri

app = Flask(__name__)
app.config['DEBUG'] = False
app.config['MONGODB_SETTINGS'] = {'host': get_db_uri(), 'connect': False}
db = MongoEngine()
db.init_app(app)
analyzer = BlockchainAnalyzer()

period_block_map = {
    'hour': 6,
    'day': 144,
    'week': 1008,
    'month': 4320
}


@app.before_request
def before_request():
    analyzer.set_blockchain(get_blockchain())
    analyzer.set_addresses(get_addresses())
    analyzer.set_utxos(get_utxos())


def _crop_params(interval, lower_height, upper_height):
    if interval < 1:
        interval = 1
    elif interval > 4320:
        interval = 4320

    if lower_height < 1:
        lower_height = 1
    elif lower_height > analyzer.max_block_number():
        lower_height = analyzer.max_block_number()

    if upper_height < lower_height:
        upper_height = lower_height
    elif upper_height > analyzer.max_block_number():
        upper_height = analyzer.max_block_number()

    return interval, lower_height, upper_height


def period_to_block_range_and_interval(period):
    upper_height = analyzer.max_block_number()

    if period == 'week': return _crop_params(period_block_map['hour'], upper_height - period_block_map['week'],
                                             upper_height)
    if period == 'month': return _crop_params(period_block_map['day'], upper_height - period_block_map['month'],
                                              upper_height)

    return _crop_params(period_block_map['week'], 1, upper_height)


@app.route('/miningdifficulty', methods=['GET'])
def mining_difficulty():
    period = request.args.get('period', default='week')
    interval, lower_height, upper_height = period_to_block_range_and_interval(period)
    difficulty = analyzer.mining_difficulty(interval, lower_height, upper_height)
    return jsonify(difficulty)


@app.route('/transactionvolume', methods=['GET'])
def transaction_volume():
    period = request.args.get('period', default='week')
    interval, lower_height, upper_height = period_to_block_range_and_interval(period)
    volume = analyzer.transaction_volume(interval, lower_height, upper_height)
    return jsonify(volume)


@app.route('/nonemptywalletsnumber', methods=['GET'])
def non_empty_wallets():
    wallets_number = analyzer.non_empty_wallets_number()
    return jsonify(wallets_number)


@app.route('/piechartdata', methods=['GET'])
def piechart_data():
    piechart_data = analyzer.piechart_data(100, 1, analyzer.max_block_number())
    return jsonify(piechart_data)


@app.route('/richestwallets', methods=['GET'])
def richest_wallets():
    limit = request.args.get('limit', default='20')
    try:
        limit = int(limit)
    except (ValueError, TypeError):
        limit = 20
    wallets = analyzer.richest_wallets(limit)
    return jsonify(wallets)


@app.route('/status', methods=['GET'])
def status():
    status = analyzer.status()
    return jsonify(status)


if __name__ == '__main__':
    app.run()
